### ISSUE TITLE
jQuery.proxy: Bring it back

### ISSUE THREAD

**Issue by Ronef** [jquery/jquery#5244]
The utility was deprecated in 3.3 (#2958).

But I don't understand, why? @mgol [wrote](https://github.com/jquery/jquery/issues/2958#issuecomment-211433159) an important thing:
> One thing that jQuery.proxy does and Function#bind doesn't is setting guid on the bound function instances so that proxied event handlers can be removed without storing the instance of the proxied function.

OK, Function#bind is faster, but is it really significant today? I ran a combat test on my calculator and saw no difference.

At the same time, it was very comfortable to remove handlers assigned with jQuery.proxy. Without it you have to write some extra code which is especially difficult for juniors. IMHO the loss is much greater than the performance gain.

**Comment by mgol** [comment]
Thanks for the report. There are some differences between `jQuery.proxy` and `Function#bind` but we wanted to encourage people to use native versions as they're close enough. We've applied this strategy to a number of other utils in the past few years.

The jQuery API obscures the fact that it's actually returning a new function because of this `guid`-based association, I think it's better to be explicit here and teach users about identity.

With regards to usage, the main difference is you need to bind beforehand and assign the result to a variable, and use that.

Also, with the event refactor we're planning for jQuery 5, we'll start using individual `addEventListener` calls per each `on( eventName, handler )` use. That may make it harder for jQuery to rely on `guid`s to find a matching handler when using `removeEventListener`. There may be ways around it but I don't want to overpromise this early. It's the safest to recommend not using this API anymore.

That said, I'll leave it open for a team discussion for a while.

**Comment by timmywil** [comment]
I think mgol's analysis is spot-on. The `guid`-based association is not future-proof.

**Comment by Ronef** [comment]
@mgol
Of course, we can save a `bind` result to a variable and use it later to remove handler.

But for example it is very inconvenient when you have a lot of class methods that need to be used in this way. As a result, you have to create an extra property for each method. Just for handlers links. A lot of unnecessary code in _my_ projects for the sake of some insignificant resource saving or bright abstract principles of people teaching :) Sorry, but clean projects code is a higher priority for me.

Another point is interaction with junior programmers. Using of jQuery.proxy has saved a lot of time. Just try googling the subject and you will see for yourself.

Save `jQuery.proxy` ≠ cancel `bind`. These are _different_ things, let all the flowers bloom. IMHO all you need to do is make the good documentation explaining the difference and outlining the scope of application of either method.

**Comment by mgol** [comment]
You ignored my remarks about planned jQuery 5 changes. Supporting `guid`-based function matching for handler removal requires jQuery to maintain an internal set of all handlers attached for element + event pairs. This is necessary now because jQuery attaches a single native handler for each such pair and traverses the handlers list by itself. In order to support `EventListenerOptions`, including adding support for passive listeners (#2871), we need to stop doing that and start using individual `addEventListener` calls. At this point, it may turn out all these internal handler storage may only be needed to handle the `guid` associations. I’m not sure it’s worth it, but we’ll only see when we get to it.

**Comment by Ronef** [comment]
@mgol
Sorry, I heard you, but i have nothing to say about it, it's just beyond me, because I'm not well-versed in internal jQuery nuances.

Thanks for your feedback. I hope all turns out well 🙏

**Comment by mgol** [comment]
Cool! For the jQuery 5 issue I mentioned alone, I'd like to leave `jQuery.proxy` as deprecated. We can revisit later. That said, if this turns out to not add too much additional complexity, I still feel a bit uneasy about it. While I understand the `guid`-based association can be helpful, `jQuery.proxy` has some other obsolete cruft like [`jQuery.proxy( context, name )`](https://api.jquery.com/jQuery.proxy/#jQuery-proxy-context-name) that I would not want to deprecate.

In any case, we're only getting close to releasing jQuery 4.0, jQuery 5.0 is a distant future at the moment. I hope the time span between 4.0 & 5.0 will be much smaller than between 3.0 & 4.0 but even if that's the case, it will still be significant time before we can seriously reconsider this deprecation.

**Comment by timmywil** [comment]
Thank you for the issue. It is unlikely that we'll be undeprecating jQuery.proxy, but a new issue can be opened post-5.0 if necessary.
